### PR TITLE
SNAP-1816 Security in Smart Connector mode

### PR DESCRIPTION
**Pull request by ashetkar**
## Changes proposed in this pull request
- Enable authentication and authz of smart connector users.
- User needs to specify 'spark.snappydata.store.user' and 'spark.snappydata.store.password' while submitting the spark job or connecting via spark-shell.

## Patch testing
precheckin in-progress

## ReleaseNotes.txt changes
NA

## Other PRs 
https://github.com/SnappyDataInc/snappy-store/pull/244

### CODE REVIEW

**Comment by kneeraj** [comment]
A general comment about GRANT/REVOKE -- Was discussing with few @vivekwiz and @dshirish that grant and revoke should ideally clear all the cached plans across sessions on lead node. Later we can make it better by invalidating only for the user for which revoked has been issued.
Right now the code is such that for row tables it is done at a place where it will be authorized only once and later a cached plan will be used. I think we can continue with this and instead clear the plan on revoke/grant and similar for column table.

File: cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+      verifyQuery1(qry1, rs1)
+      rs1.close()
+      // Thread.sleep(1000000)
+    } finally {
+      stmt1.close()
+      conn.close()
+    }
+  }
+
+  def createTable1(serverHostPort: Int, tableName: String, user: String, pass: String): Unit = {
```
Comment (kneeraj) [inline cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala]: Instead of names like 1 and 2 appended to method names we should use a little informative names otherwise you need to look at the code to figure out the difference between 1 and 2 etc.

File: cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    val stmt1 = conn.createStatement()
+    try {
+      stmt1.execute(s"drop table $tableName")
+    } finally {
+      stmt1.close()
+      conn.close()
+    }
+  }
+
+  def createTable2(serverHostPort: Int, tableName: String, user: String, pass: String): Unit = {
```
Comment (kneeraj) [inline cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala]: See above comment on naming.

File: core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+  val jdbcUser1 = "gemfire1"
+  val jdbcUser2 = "gemfire2"
+  val adminUser1 = "gemfire3"
+
+  override def setUp(): Unit = {
+    super.setUp()
+  }
+
+  override def tearDown2(): Unit = {
+    super.tearDown2()
```
Comment (dshirish) [inline core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala]: Need to stop the ldap server as done in the other test?

Comment (dshirish) [inline core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala]: Ok done in the afterclass

File: cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+  
+  def testColumnTableRouting(): Unit = {
+    val serverHostPort = AvailablePortHelper.getRandomAvailableTCPPort
+    vm2.invoke(classOf[ClusterManagerTestBase], "startNetServer", serverHostPort)
+    // scalastyle:off println
+    println(s"network server started at $serverHostPort")
+    // scalastyle:on println
+
+    val tableName = "order_line_col"
+    createTable1(serverHostPort, tableName, jdbcUser, jdbcUser)
```
Comment (kneeraj) [inline cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala]: Where is the negative testing. Like a user who is not authorized to touch tabl1 created by jdbcUser

File: cluster/src/dunit/scala/io/snappydata/cluster/SplitSnappyClusterDUnitTest.scala [hunk 5]
```diff
@@ -342,8 +342,7 @@ object SplitSnappyClusterDUnitTest
       prop: Properties,
       locatorClientPort: Int): Unit = {
 
-    val snc: SnappyContext = getSnappyContextForConnector(locatorPort,
-      locatorClientPort)
+    val snc: SnappyContext = getSnappyContextForConnector(locatorClientPort)
```
Comment (kneeraj) [inline cluster/src/dunit/scala/io/snappydata/cluster/SplitSnappyClusterDUnitTest.scala]: Please explain what are these changes? of locatorPort and locatorClientPort.

Only locatorClientPort should suffice and thats the change I can see. But what was locatorPort here?

File: core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+  private def writeToFile(str: String, fileName: String): Unit = {
+    val pw = new PrintWriter(fileName)
+    try {
+      pw.write(str)
+    } finally {
+      pw.close()
+    }
+  }
+
+  override def testColumnTableCreation(): Unit = {}
```
Comment (kneeraj) [inline core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala]: @ashetkar these are the ones you said you are completing. Right?

File: core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+
+  override def testRowTableCreation(): Unit = {}
```
Comment (kneeraj) [inline core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala]: same as above.

File: core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+
+  override def _testComplexTypesForColumnTables_SNAP643(): Unit = {}
+
+  def _testTableFormChanges(): Unit = {}
```
Comment (kneeraj) [inline core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala]: You want to get rid of this?

File: core/src/main/scala/org/apache/spark/sql/hive/HiveClientUtil.scala [hunk 9]
```diff
@@ -149,19 +156,23 @@ class HiveClientUtil(val sparkContext: SparkContext) extends Logging {
...
-        dbURL + ";default-schema=" + Misc.SNAPPY_HIVE_METASTORE + ";"
+        logURL = dbURL + ";default-schema=" + Misc.SNAPPY_HIVE_METASTORE + ";user=" + user.get
+        logURL + ";password=" + password.get + ";"
       } else {
         metadataConf.setVar(HiveConf.ConfVars.METASTORE_CONNECTION_USER_NAME,
           Misc.SNAPPY_HIVE_METASTORE)
         dbURL
       }
-      logInfo(s"Using SnappyStore as metastore database, dbURL = $secureDbURL")
+      logInfo(s"Using SnappyStore as metastore database, dbURL = $logURL")
```
Comment (dshirish) [inline core/src/main/scala/org/apache/spark/sql/hive/HiveClientUtil.scala]: Need to mask the password.

File: core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala [hunk 10]
```diff
@@ -99,12 +100,22 @@ class SnappyStoreHiveCatalog(externalCatalog: SnappyExternalCatalog,
...
     val defaultDbDefinition =
       CatalogDatabase(defaultName, "app database", sqlConf.warehousePath, Map())
-    // Initialize default database if it doesn't already exist
-    externalCatalog.createDatabase(defaultDbDefinition, ignoreIfExists = true)
-    client.setCurrentDatabase(defaultName)
+    SnappyContext.getClusterMode(snappySession.sparkContext) match {
+      case ThinClientConnectorMode(_, _) =>
+      case _ =>     // Initialize default database if it doesn't already exist
+        externalCatalog.createDatabase(defaultDbDefinition, ignoreIfExists = true)
+        client.setCurrentDatabase(defaultName)
```
Comment (dshirish) [inline core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala]: Just to confirm- even the call to "client.setCurrentDatabase(defaultName)" needs to be skipped in connector mode and not just the createDatabase() call?

File: core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala [hunk 11]
```diff
@@ -82,7 +83,7 @@ class SnappyStoreHiveCatalog(externalCatalog: SnappyExternalCatalog,
 
   val sparkConf: SparkConf = snappySession.sparkContext.getConf
 
-  private[sql] var client = metadataHive
+  private[sql] var client = metadataHive // externalCatalog.client
```
Comment (kneeraj) [inline core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala]: you want to remove the commented part?

File: core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala [hunk 12]
```diff
@@ -99,12 +100,22 @@ class SnappyStoreHiveCatalog(externalCatalog: SnappyExternalCatalog,
...
       CatalogDatabase(defaultName, "app database", sqlConf.warehousePath, Map())
-    // Initialize default database if it doesn't already exist
-    externalCatalog.createDatabase(defaultDbDefinition, ignoreIfExists = true)
-    client.setCurrentDatabase(defaultName)
+    SnappyContext.getClusterMode(snappySession.sparkContext) match {
+      case ThinClientConnectorMode(_, _) =>
+      case _ =>     // Initialize default database if it doesn't already exist
+        externalCatalog.createDatabase(defaultDbDefinition, ignoreIfExists = true)
+        client.setCurrentDatabase(defaultName)
+    }
```
Comment (dshirish) [inline core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala]: Probably the initialization code of currentSchema can be arranged differently
In case of ThinClientConnectorMode(_, _),  get user as required by using Constant.SPARK_STORE_PREFIX property in the other case get it using just Attribute.USERNAME_ATTR property. 
Also in the current method formatDatabaseName() gets called twice if the 'user' is set (that is not Constant.DEFAULT_SCHEMA)

File: core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+    vm2 = host.getVM(2)
+    vm3 = host.getVM(3)
+  }
+
+  val jdbcUser1 = "gemfire1"
+  val jdbcUser2 = "gemfire2"
+  val adminUser1 = "gemfire3"
+
+  override def setUp(): Unit = {
+    super.setUp()
```
Comment (dshirish) [inline core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala]: Probably can skip override of setUp and tearDown2

File: core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+    assert(rows == 1005, s"expected 1005 rows but found $rows in $eRowTab1")
+
+    // drop
+    snc.sql(s"drop table $eColTab1")
+    snc.sql(s"drop table $eRowTab1")
+    assert(!snc.sparkSession.catalog.tableExists(eColTab1), s"$eColTab1 not dropped")
+    assert(!snc.sparkSession.catalog.tableExists(eRowTab1), s"$eRowTab1 not dropped")
+
+    // vice versa
+  }
```
Comment (dshirish) [inline core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala]: Please add test case where user tries to update tables in hive metastore and it should fail.
Also test cases for alter table scenario.

File: core/src/main/scala/org/apache/spark/sql/hive/HiveClientUtil.scala (hunk 9 above)
Comment (ashetkar) [inline core/src/main/scala/org/apache/spark/sql/hive/HiveClientUtil.scala]: logURL doesn't contain the password, secureURL does.

File: core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala (hunk 11 above)
Comment (ashetkar) [inline core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala]: Was wondering why we are passing metadataHive param separately. It's already accessible as externalCatalog.client. Was thinking to replace it but may change a few signatures.

File: core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala (hunk 10 above)
Comment (ashetkar) [inline core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala]: Yes, as it too tries to write something into the hive metastore and throws error (don't recollect exact error)

File: cluster/src/dunit/scala/io/snappydata/cluster/SplitSnappyClusterDUnitTest.scala (hunk 5 above)
Comment (ashetkar) [inline cluster/src/dunit/scala/io/snappydata/cluster/SplitSnappyClusterDUnitTest.scala]: locatorPort was for locator discovery by peers which was being used earlier in split cluster mode. As we have moved to smart connector as a thin client, this was redundant.

File: cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+  val jdbcUser = "gemfire1"
+
+  override def setUp(): Unit = {
+    setSecurityProps()
+    super.setUp()
+  }
+
+  override def tearDown2(): Unit = {
+    super.tearDown2()
+    val ldapServer = LdapTestServer.getInstance()
```
Comment (ashetkar) [inline cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala]: This query routing dunit file is not part of this PR. It got in here due to a merge went awry. Will update.

File: cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+import com.pivotal.gemfirexd.security.{LdapTestServer, SecurityTestUtils}
+import io.snappydata.test.dunit.AvailablePortHelper
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.collection.Utils
+
+class SecureQueryRoutingDUnitTest(val s: String)
+    extends ClusterManagerTestBase(s) with Logging {
+
+  val snappyAdminUser = "gemfire2"
```
Comment (ashetkar) [inline cluster/src/dunit/scala/io/snappydata/cluster/SecureQueryRoutingDUnitTest.scala]: This query routing dunit file is not part of this PR. It got in here due to a merge went awry. Will update.

File: core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala (hunk 8 above)
Comment (ashetkar) [inline core/src/dunit/scala/io/snappydata/cluster/SecureSplitClusterDUnitTest.scala]: Should keep it without _, to avoid running its parent's implementation.

**Comment by ashetkar** [comment]
This PR has been merged into SNAP-1656.
A separate PR will be opened for merge into master.

File: cluster/src/dunit/scala/io/snappydata/cluster/SplitSnappyClusterDUnitTest.scala (hunk 5 above)
Comment (kneeraj) [inline cluster/src/dunit/scala/io/snappydata/cluster/SplitSnappyClusterDUnitTest.scala]: Yes, thought so but I thought this would have removed after @dshirish removed his changes of old code. Ok.
